### PR TITLE
Add mbedtls_ssl_ticket_rotate() for ticket rotation

### PR DESCRIPTION
**Pull request by gstrauss**
## Description
Add `mbedtls_ssl_ticket_rotate()` for ticket rotation
#5421 Add access to some fields of `mbedtls_ssl_ticket`

## Status
**READY**

## Requires Backporting
NO  

## Migrations
NO

## Todos
- [x] Tests
- [x] Documentation
- [x] Changelog updated

### CODE REVIEW

**Comment by gstrauss** [comment]
FYI: `mbedtls_ssl_ticket_rotate()` fails for any reason, the default internal session ticket encryption key rotation remains in effect.  If using `mbedtls_ssl_ticket_rotate()` to provide session ticket encryption keys across different machines, `mbedtls_ssl_ticket_rotate()` must be called at least as frequently as the ticket lifetime, or else the default internal session ticket encryption key rotation will rotate the keys and they will no longer be synchronized across machines.  For security reasons, the keys continue to be rotated internally even if there is a failure to push new keys externally.  The impact in this scenario would be less efficient connection initiation due to failed session resumption when connecting to different machines with different session ticket encryption keys.  However, from a security stance, this reduced efficiency is preferred to excessively long lifetimes of session ticket encryption keys.

File: library/ssl_ticket.c
Comment (mpg) [inline library/ssl_ticket.c:125]: s/ticket/key/ I guess?

File: include/mbedtls/ssl_ticket.h
Comment (mpg) [inline include/mbedtls/ssl_ticket.h:119]: Style: s/len/length/ (avoid abbreviations in user documentation) and add "in bytes". (Same for klen below, especially important there are key sizes in bits are common).

File: include/mbedtls/ssl_ticket.h
Comment (mpg) [inline include/mbedtls/ssl_ticket.h:125]: Thanks for documenting that. Perhaps also mention that `klen` must not be less than the previous value?

**Review by mpg — CHANGES_REQUESTED** [review]
Looks good to me except for minor documentation nits. Thanks for providing a high-level API that meets your need without requiring direct field access.

**Comment by mpg** [comment]
> If using `mbedtls_ssl_ticket_rotate()` to provide session ticket encryption keys across different machines, `mbedtls_ssl_ticket_rotate()` must be called at least as frequently as the ticket lifetime, or else the default internal session ticket encryption key rotation will rotate the keys and they will no longer be synchronized across machines.

Actually, could add something like that to the documentation as well?

Regarding testing, I think full testing would be pretty hard (we'd need several server instances), so I'd say we can skip it. OTOH, I don't think I'm comfortable with the function never being executed at all in our test suite either. So, I was thinking perhaps just add a command-line option like `ticket_external_key` to `programs/ssl/ssl_server2.c` that would cause `mbedtls_ssl_ticket_rotate()` to be called once at startup (say, right after `mbedtls_ssl_ticket_setup()`), with a comment explaining that this is how you would synchronize keys across multiple instances, and that you'd have to renew the call often enough to keep keys synchonized. That's very minimal (basically it just ensures that the function doesn't crash and leaves the ticket context in a usable state), but better than nothing, and it also completes the documentation with an example, however incomplete. Unless of course you can think of something better. Wdyt?

**Comment by gstrauss** [comment]
Doc has been updated with your suggestions and further explanations have been added.

Still missing test coverage.

**Comment by gstrauss** [comment]
> Regarding testing, I think full testing would be pretty hard (we'd need several server instances), so I'd say we can skip it. OTOH, I don't think I'm comfortable with the function never being executed at all in our test suite either. So, I was thinking perhaps just add a command-line option like `ticket_external_key` to `programs/ssl/ssl_server2.c` that would cause `mbedtls_ssl_ticket_rotate()` to be called once at startup (say, right after `mbedtls_ssl_ticket_setup()`), with a comment explaining that this is how you would synchronize keys across multiple instances, and that you'd have to renew the call often enough to keep keys synchonized.

Additional commit added which attempts this (and passes tests).  It can be adjusted as you see fit.

**Comment by gstrauss** [comment]
The test I modified to exercise `mbedtls_ssl_ticket_rotate()` is failing in github CI, but the test passes for me when I run it on my local machine.  The code I added to `programs/ssl/ssl_server2.c` will print out an error message with the error code.  Is that trace in another log that I can see from the github CI?  (SERVER START TIMEOUT is what I see if I force `mbedtls_ssl_ticket_rotate()` to fail with this test on my local machine)
```
Session resume using tickets: basic .................................... SERVER START TIMEOUT
tests/ssl-opt.sh: 1091: kill: No such process
FAIL
  ! server or client failed to reach handshake stage
  ! outputs saved to o-XXX-123.log
```

**Review by mpg — COMMENTED** [review]
Everything looks good to me, except the CI failure.

**Comment by mpg** [comment]
I was able to reproduce the CI failure locally with: `scripts/config.py full && CC=gcc cmake -D CMAKE_BUILD_TYPE:String=Asan . && make ssl_client2 ssl_server2 udp_proxy query_compile_time_config && tests/ssl-opt.sh -f 'using tickets: basic'`. Turns out what makes it fail is `MBEDTLS_USE_PSA_CRYPTO` is set, causing `mbedtls_cipher_setup_psa()` to be used in `mbedtls_ssl_ticket_setup()`, and then `mbedtls_cipher_set_key()` can [no longer be called more than once](https://github.com/ARMmbed/mbedtls/blob/development/library/cipher.c#L281).

So, unless I'm mistaken, that means automatic key rotation is also broken with `MBEDTLS_USE_PSA_CRYPTO`, so we have a pre-existing bug and a corresponding testing gap. Sigh.

I'll think about it a bit and come back to you regarding what we can do. We need to consider (1) that in development we're soon revamping how `MBEDTLS_USE_PSA_CRYPTO` is implemented in `ssl_ticket.c` and moving away from `mbedtls_cipher_setup_psa()` - #5203 - which hopefully can fix the bug naturally, and (2) that the bug needs to be fixed in 2.28 too.

**Comment by mpg** [comment]
Ok, after thinking about it and discussing with a colleague, I think the easiest thing to do is to temporarily accept that `mbedtls_ssl_ticket_rotate()` is incompatible with `MBEDTLS_USE_PSA_CRYPTO`, then we'll fix that incompatibility as part of #5203 - which is going to be started soon.

In the meantime, I suggest the following strategy:
- in `tests/ssl-opt.sh` instead of re-using the existing `tickets: basic` test, leave it untouched, and add a new test below `tickets: manual rotation` (I was already on the fence about recommending that anyway)
- add `requires_config_disabled MBEDTLS_USE_PSA_CRYPTO` before this new test - this part is temporary and we'll remove it in #5203 

That should allow this PR to be progress without being blocked by a pre-existing bug, that just happened to be revealed by the new test.

**Comment by gstrauss** [comment]
I modified the test as recommended above.  It PASSes in my default build environment.  Using your much quicker, targeted build steps (thank you), I can confirm that the *new* test is SKIPped, whereas the 'basic' test would FAIL before the change.

**Review by mpg — APPROVED** [review]
LGTM, thanks for fixing the tests as suggested. Now waiting for a second reviewer.

File: programs/ssl/ssl_server2.c
Comment (mprse) [inline programs/ssl/ssl_server2.c:2755]: I'm not sure if defining macro inside if statement is consistent with the codding standards. At least for readability, a preprocessor directive should begin in the first column.

Maybe this can be changed to:
`const int max_key_bytes = 32;`

Comment (mpg) [reply programs/ssl/ssl_server2.c:2755]: Actually, I think this points to a larger problem: other users of this function are likely to need the same constants, and having everyone look at the implementation in `ssl_ticket.c` is not ideal.

Now I'm thinking we should probably export macros `MBEDTLS_SSL_TICKET_MAX_KEY_BYTES` and `MBEDTLS_SSL_TICKET_KEY_NAME_BYTES` in `ssl_ticket.h` (and use them in `ssl_ticket.c` instead of the currently-internal versions).

Should have noticed earlier, a comment that references a `.c` file is often a sign that the public API or documentation is not good enough. (That's another reason why writing tests is useful, helps notice this kind of thing.)

**Review by mpg — CHANGES_REQUESTED** [review]
Changed my mind since my previous approval, now I'd like the public API to be extended with suitable constants.

**Comment by gstrauss** [comment]
> Now I'm thinking we should probably export macros `MBEDTLS_SSL_TICKET_MAX_KEY_BYTES` and `MBEDTLS_SSL_TICKET_KEY_NAME_BYTES` in ssl_ticket.h

I added a commit defining these in ssl_ticket.h

**Review by mpg — APPROVED** [review]
LGTM

**Review by mprse — APPROVED** [review]

**Comment by mpg** [comment]
@gstrauss Coming back to this, it's not really clear to me why `mbedtls_ssl_ticket_rotate()` needs a `lifetime` parameter that updates the context's lifetime field. I'd expect we pick a ticket lifetime when calling `mbedtls_ssl_ticket_setup()` and then stick with it for as long as we use this context. Can you clarify why you made `rotate()` also update `ctx->ticket_lifetime`? Thanks in advance!

Cc @ronald-cron-arm

**Comment by gstrauss** [comment]
> @gstrauss Coming back to this, it's not really clear to me why `mbedtls_ssl_ticket_rotate()` needs a `lifetime` parameter that updates the context's lifetime field. I'd expect we pick a ticket lifetime when calling `mbedtls_ssl_ticket_setup()` and then stick with it for as long as we use this context. Can you clarify why you made `rotate()` also update `ctx->ticket_lifetime`?

tl;dr: custom management of the session ticket encryption key

lighttpd mod_mbedtls provides an option to maintain the session ticket encryption key *externally* from lighttpd, so that the key can be shared by multiple instances of lighttpd across multiple servers, as long as an admin configures a process to synchronize lighttpd.conf `ssl.stek-file`.

https://git.lighttpd.net/lighttpd/lighttpd1.4/src/branch/master/src/mod_mbedtls.c

When the STEK is maintained outside from lighttpd, lighttpd allows the admin to specify the ticket lifetime, and lighttpd mod_mbedtls sets that lifetime in `mbedtls_ssl_ticket_context` when managing the STEK, to communicate the lifetime into the mbedtls library, separate from the mbedtls library defaults when the mbedtls library manages the session ticket encryption key rotation.

**Comment by mpg** [comment]
Thanks for your prompt reply! I think I understand the general principle of custom management of the STEKs, but even after looking at the code, I don't understand why `rotate()` needs a `lifetime` argument that updates `ctx->ticket_lifetime`.

Imagine we wanted to remove that argument in 4.0, making the new signature:
```c
int mbedtls_ssl_ticket_rotate(mbedtls_ssl_ticket_context *ctx,
                              const unsigned char *name, size_t nlength,
                              const unsigned char *k, size_t klength);
```
Could you explain what concrete problem that would cause with your use case?

(Sorry if this is obvious. Perhaps after you explained I'll go "oh right of course", but right now I'm not seeing it.)

**Comment by gstrauss** [comment]
https://git.lighttpd.net/lighttpd/lighttpd1.4/src/branch/master/src/mod_mbedtls.c#L363

`ctx->ticket_lifetime` is a private member in mbedtls 3.x.  For me to set a custom lifetime (shorter than the mbedtls default), I need access to set this value.  That is why `mbedtls_ssl_ticket_rotate()` has that parameter.

This PR addresses #5421
As you wrote in #5421 
> However, in that instance it looks like the application only wants to slightly extend key management, so writing a full replacement to ssl_ticket.c would be a bit much for the task at hand, and it is desirable to allow applications to re-use most of ssl_ticket.c but just modifying the bits they need by accessing struct fields.

**Comment by mpg** [comment]
Thanks for your response! 



> `ctx->ticket_lifetime` is a private member in mbedtls 3.x. For me to set a custom lifetime (shorter than the mbedtls default), I need access to set this value. That is why `mbedtls_ssl_ticket_rotate()` has that parameter.

Here's the point I'm not getting: you can already set that value using `mbedtls_ssl_ticket_setup()`, and looking at your code, I can see that you do (lines 1567/1573 from the link above). Intuitively, I'd expect this is a value that's going to be set once and never changed later. Apparently my intuition is not correct since you're feeling the need to update that value when calling `ticket_rotate()`, but I'd like to understand why.

> If the mbedtls team needs to make changes here, please help me understand the goals. Are the changes aimed at TLS 1.3 session tickets?
> 
> While I would like to avoid having to fork and maintain ssl_ticket.c as a "custom implementation" -- identical to ssl_ticket.c except for access to set `ctx->ticket_lifetime` -- I can do that, but would like to understand why the changes are needed. Thanks.

Sorry, I think the wording of my previous message came out to strong: we won't remove that parameter if you really need it, and we don't want users to have to maintain a fork of `ssl_ticket.c` if we can avoid it. The reason I'm asking is precisely so we can avoid that!

The _potential_ changes are indeed related to TLS 1.3. It came up during discussion that our implementation could be slightly simpler if `ctx->ticket_lifetime` were set once (by `setup()`) and never changed later, which made us notice `rotate()` is the only function that can modify it after `setup()`. However, if that's needed, then we can go with a slightly more complex design that supports mutating `ctx->ticket_lifetime` - we'd just like to make sure this is actually needed, by understanding why.

**Comment by gstrauss** [comment]
> The _potential_ changes are indeed related to TLS 1.3. It came up during discussion that our implementation could be slightly simpler if `ctx->ticket_lifetime` were set once (by `setup()`) and never changed later, which made us notice `rotate()` is the only function that can modify it after `setup()`. However, if that's needed, then we can go with a slightly more complex design that supports mutating `ctx->ticket_lifetime` - we'd just like to make sure this is actually needed, by understanding why.

Practically speaking, lighttpd mod_mbedtls has not tried to read an external STEK file at the time that `mbedtls_ssl_ticket_setup()` is called, and that external STEK file is an optional feature of lighttpd mod_mbedtls, which may or may not be configured.  lighttpd calls `mbedtls_ssl_ticket_setup()` at lighttpd startup, and *if* an external STEK file is configured and exists, then that is handled later and separately.  Crucially, if the external STEK file is configured but (temporarily?) does not exist, then session tickets still work with the mbedtls library defaults, and will still be rotated periodically as is good security practice, but the session ticket encryption key is not shared across multiple independent instances, so session resumption may be less efficient if clients try to resume sessions with other independent lighttpd instances.  If the STEK file later comes into existence, lighttpd will read and honor the ticket lifetime set in the external STEK file.  Maintaining this external STEK file, if configured in lighttpd, is done by the admin using an external process *independent from lighttpd*, and there can be race conditions synchronizing the file across multiple machines, e.g. VMs are spun up and processes are started.  For example, temporary inability to reach out to an external server to obtain the shared STEK file should not block startup of lighttpd, since TLS will work fine even if session resumption attempts do not succeed.

I hope this explains some reasons how `mbedtls_ssl_ticket_rotate()` might be used to set ticket lifetime after `mbedtls_ssl_ticket_setup()`.

While I could modify lighttpd mod_mbedtls to check for the STEK file prior to `mbedtls_ssl_ticket_setup()`, and I do not expect an admin to change the ticket lifetime security policy in the external STEK file once the STEK file is read, I would still like a user-friendly and security-aware way to recover and honor the security policy in the external STEK file if the external STEK file is not present at lighttpd startup, but later does exist.  A separate accessor function to change ticket lifetime, independent from `mbedtls_ssl_ticket_rotate()` would be fine -- or something else you might suggest -- though when I would need to set this information is about the same time I call `mbedtls_ssl_ticket_rotate()`.

Conjecturing about how you are trying to simplify `mbedtls_ssl_ticket_rotate()`, I would also be okay with `mbedtls_ssl_ticket_rotate()` allowing UINT32_MAX for the ticket lifetime to mean "no change" from the value set when `mbedtls_ssl_ticket_setup()` was called, or passing in `(uint32_t *)`, so that if NULL is passed, it means "no change".

**Comment by mpg** [comment]
> Practically speaking, lighttpd mod_mbedtls has not tried to read an external STEK file at the time that `mbedtls_ssl_ticket_setup()` is called, and that external STEK file is an optional feature [...] If the STEK file later comes into existence, lighttpd will read and honor the ticket lifetime set in the external STEK file.
> 
> I hope this explains some reasons how `mbedtls_ssl_ticket_rotate()` might be used to set ticket lifetime after `mbedtls_ssl_ticket_setup()`.
> 
I does, thanks for the detailed explanation, I think it completely makes sense to me now!

> While I could modify lighttpd mod_mbedtls to check for the STEK file prior to `mbedtls_ssl_ticket_setup()`, and I do not expect an admin to change the ticket lifetime security policy in the external STEK file once the STEK file is read, I would still like a user-friendly and security-aware way to recover and honor the security policy in the external STEK file if the external STEK file is not present at lighttpd startup, but later does exist. A separate accessor function to change ticket lifetime, independent from `mbedtls_ssl_ticket_rotate()` would be fine -- or something else you might suggest -- though when I would need to set this information is about the same time I call `mbedtls_ssl_ticket_rotate()`.
> 
Well actually for us it doesn't really matter which function is used to update the lifetime, the distinction is about whether the lifetime can change at all. Now I understand why it may be necessary for the lifetime to change, and handling this is no big deal for us, so I don't think we'll be doing changes to that part of the API.

Thanks again for taking the time to explain. Having this kind of feedback from users is extremely helpful.

**Pull request merged**
